src/EnergyPlus/SurfaceOctree.cc

```diff
@@ -0,0 +1,265 @@
+// You are under no obligation whatsoever to provide any bug fixes, patches, or upgrades to the
+// features, functionality or performance of the source code ("Enhancements") to anyone; however,
+// if you choose to make your Enhancements available either publicly, or directly to Lawrence
+// Berkeley National Laboratory, without imposing a separate written license agreement for such
+// Enhancements, then you hereby grant the following license: a non-exclusive, royalty-free
+// perpetual license to install, use, modify, prepare derivative works, incorporate into other
+// computer software, distribute, and sublicense such enhancements or derivative works thereof,
+// in binary and source code form.
+
+// EnergyPlus Headers
+#include <EnergyPlus/SurfaceOctree.hh>
+#include <EnergyPlus/DataSurfaces.hh>
+
+// C++ Headers
+#include <algorithm>
+#include <cmath>
+#include <limits>
+
+namespace EnergyPlus {
+
+// Package: Surface Octree System
+//
+// Purpose: Spatial sort of surfaces for fast, scalable identification of active surfaces for algorithms
+//  making spatial queries such as solar shading, solar reflection, and daylighting obstruction
+//
+// Author: Stuart Mentzer (Stuart_Mentzer@objexx.com)
+//
+// History:
+//  Sep 2015: Experimental code
+//  Jan 2016: Initial release
+//
+// Notes:
+//  Initial octree is for use in daylighting but should be adaptable to other use cases:
+//   Surfaces without vertices are omitted
+//   Transparent surfaces are omitted (can't obstruct light)
+//  Copy and move ctors/assignment omitted for now since not needed
+//  The use of multiple octrees for faster lookups of surface type subsets may be worthwhile for performance in some uses
+//  Octree variations and parameter tuning can give better performance for specific applications
+//  This design uses "tight" cubes (no overlap) and surfaces filtering down to deepest cube they fit in completely
+//  Alternative: Use "loose" cubes oversied by x2 or some other factor to allow surfaces to filter down further: This requires more cubes to be processed for a given operation
+//  Alternative: Filter all surfaces down to leaf cubes placing a surface in any cube it intersects: More specificity but redundant surfaces in each operation so must collect them in a set
+
+	// Surface in Cube?
+	bool
+	SurfaceOctreeCube::
+	contains( Surface const & surface ) const
+	{
+		for ( Vertex const & v : surface.Vertex ) { // All surface vertices must be in cube
+			if ( ! contains( v ) ) return false;
+		}
+		return true;
+	}
+
+	// Surfaces Outer Cube Initilization
+	void
+	SurfaceOctreeCube::
+	init( ObjexxFCL::Array1< Surface > & surfaces )
+	{
+		assert( d_ == 0u );
+		assert( n_ == 0u );
+		surfaces_.clear();
+		surfaces_.reserve( surfaces.size() );
+		for ( Surface & surface : surfaces ) {
+			if (
+			 ( surface.Vertex.size() >= 3 ) && // Skip no-vertex "surfaces"
+			 ( ! surface.IsTransparent ) // Skip transparent surfaces
+			) {
+				surfaces_.push_back( &surface );
```

Comment:
@mbadams5 yes this certainly depends on Surface stability and lifetime as would any associated collection. We don't have the machinery in place to manage this all automatically -- that is miles of OO away from where we are now. We should be using various linked sub-collections in many places for code efficiency and clarity and the same issue will arise there. We could come up with an API for making changes to the Surfaces if/when that becomes something that EnergyPlus needs to do.

I will add some comments on this and some asserts to help detect changes to Surface.